Dask usage at blue yonder (#2)
andreas-merkel-by/jdatech.github.io: _posts/2020-03-16-dask-usage-at-blue-yonder.markdown

---
layout: single
title:  "Dask Usage at Blue Yonder"
date:   2020-03-16 10:00:00 +0100
tags: technology python data-engineering
header:
  overlay_image: assets/images/tech_gear_banner.jpg
  overlay_filter: 0.2
  show_overlay_excerpt: false

> author: Andreas Merkel, Kshitij Mathur

Reviewer: You need to have the entry of `Andreas Merkel, Kshitij Mathur` in `authors.yml` for this to "work", e.g. https://github.com/JDASoftwareGroup/jdatech.github.io/blob/1b4ab751c9946321272f4a918fd81db509113af7/_data/authors.yml#L71
(You can test this locally using the last command provided in the readme)

Author: Thx, I was already suspecting it's not that smart...

> ---

Reviewer 2: I know I'm a markdown Nazi, but I personally _really_ like to rely on markdown's formatter, i.e. one sentence = one line. That's IMHO much easier to review/comment then mixed sentences in one line.

```
This is a sentence.
This is another sentence.
And this is how markdown renders it.
```

This is a sentence. This is another sentence. And this is how Markdown renders it if everything is configured correctly.

Reviewer: Why not a paragraph per line? Then, what you see is what you get and it makes inserting suggestions easier

Reviewer 2: IMHO then it's pretty hard to review.

The changelog might look like 60 lines for adding a single `,`.

Reviewer 3: Or if you don't like to do sentences, then at least use a reasonable line length (like 120chars).

Author: I think that's what I did. For longer sentences, the lines get longer than I like otherwise. But it might be a good idea to start each new sentence on a new line.

layout: single
title:  "Dask Usage at Blue Yonder"
date:   2020-03-16 10:00:00 +0100
tags: technology python data-engineering
header:
  overlay_image: assets/images/tech_gear_banner.jpg
  overlay_filter: 0.2
  show_overlay_excerpt: false
author: Andreas Merkel, Kshitij Mathur
author_profile: true
---
# Dask Usage at Blue Yonder

Back in 2017, Blue Yonder started to look into 
[Dask/Distributed](https://distributed.dask.org) and how we can leverage it to power
our machine learning and data pipelines.
It's 2020 now, and we are using it heavily in production for providing demand predictions,

> recommended order quantities, and price proposals to our customers.

Reviewer 4: Not sure if an external reader actually understands what we mean with demand predictions and oder quantiles. I think we should assume that a reader does not actually know what Blue Yonder does (and even if somebody knows BY, BY is nowadays a quite large company with many applications)

Time to take a a look at the way we are using Dask!

## Use cases

First, let's have a look at the use cases we have for Dask. The use of Dask at Blue Yonder
is strongly coupled to the concept of [Datasets](https://tech.jda.com/introducing-kartothek/),

> tabular data stored as Apache [Parquet](https://parquet.apache.org/) files in blob storage. 

Reviewer 2: ```suggestion
tabular data stored as [Apache Parquet](https://parquet.apache.org/) files in blob storage. 
```

> We use these datasets for storing everything that is needed
> for generating predictions, recommended order quantities, and recommended prices for our customers,
> as well as the predictions, order quantities, and prices themselves.

Reviewer 2: I needed to read that multiple times to really understand the sentence.

1. Emphasize `storing`?
2. The enumeration of `for generating predictions, recommended order quantities, and recommended prices` looks wrong. You want to stick to one tense/format, i.e. `recommended` → `recommend` or even further to change the verb.
3. Do you mean _Dask_ and not the datasets here? I still do not understand why we are mentioning them twice... 
```suggestion
We use datasets for storing everything that is needed
for generating predictions and creating order quantity recommendations or price recommendations for our customers.
Dask is used in managing/creating this data as well as performing the necessary computations.
```

The size of the data we work on varies strongly depending on the individual customer, with the largest ones

> currently amounting to about 300 GiB of compressed Parquet data per day, corresponding to about 500 GiB of

Reviewer 4: Where do these numbers come from?
1. 300GiB a day sounds a bit too large
2. We have usually compression rates of at least 20, i.e. 300GiB compressed would amount to about ~6TB in memory.

Maybe we can cite number of rows as well? I guess num of rows is more universal, storage size and in memory depends on so many things

Author: I got it from your Dask summit slides, but out of unconscious delusions of grandeur added a zero. 😂

> The size of the data we work on varies strongly depending on the individual customer, with the largest ones
> currently amounting to about 300 GiB of compressed Parquet data per day, corresponding to about 500 GiB of
> uncompressed data in memory.

Reviewer 2: ```suggestion
The size of the data we work on varies strongly depending on the individual customer.
The largest ones currently amount to about 300 GiB of compressed Parquet data per day, corresponding to about 500 GiB of uncompressed data in memory.
```

### Downloading data from a database to a dataset

> One typical use case that we have is downloading large amounts of data from a database into a

Reviewer 4: One important bit about these graphs would be that we also need to limit concurrency (which is why we built the semaphore. We could link https://docs.dask.org/en/latest/futures.html?highlight=semaphore#id1

dataset. We use Dask/Distributed here as a means to parallelize the download and the graph is

> an extremely simple one: a lot of parallel nodes downloading pieces of data into a blob and one

Reviewer 4: The technical term for "extremely simple" graph in this case would be [embarrassingly parallel](https://en.wikipedia.org/wiki/Embarrassingly_parallel)

> reduction node writing the dataset metadata (which is basically a dictionary what data can be found in which blob).

Reviewer 4: Might be another opportunity to drop a link to kartothek

Author: I have a link to the Kartothek blog post in the section before, but I'll add one here for good measure 😄

### Dataset maintenance

We use Dask/Distributed for performing maintenance operations like garbage collection on datasets.
Garbage collection is necessary since we use a lazy approach when updating datasets and only delete or update
references to blobs in the dataset metadata, deferring the deletion of no longer used blobs.
In this case, Dask is used to parallelize the delete requests to the storage provider (Azure blob service).
In a similar fashion, we use Dask to parallelize copy operations for backing up datasets.

### Repartitioning a dataset

Our algorithms rely on source datasets that are partitioned in a particular way. It is not always
the case that we have the data available in a suitable partitioning, for instance, if the data results
from a computation that used a different partitioning. In this case, we have to repartition the data.
This works by reading the dataset as a [Dask dataframe](https://docs.dask.org/en/latest/dataframe.html) 
repartitioning the dataframe using network shuffle, and writing it out again to a dataset.

## Ad hoc data analysis

Finally, our data scientists also use Dask for ad hoc analysis of data that is too large to

> fit into an individual compute node's memory.

Reviewer 2: Should we incorporate the buzzword `out-of-core` here?

Reviewer 4: +1 for buzzword bingo

We can also mention that this happens from within a jupyter notebook (mentioning OSS technology is usually a good idea)

## Dask cluster setup at Blue Yonder

We run a somewhat unique setup of Dask clusters that is partly driven by the specific requirements

> of our domain and partly historically grown. We are running a large number of individual clusters

Reviewer 2: We run ... We are running ...?

```suggestion
of our domain and partly historically grown. We operate a large number of individual clusters
```

> of our domain and partly historically grown. We are running a large number of individual clusters

Reviewer 4: I would argue this is not historically grown but intentionally. Keyword: "environment isolation for SaaS applications" (And data isolation between customers) .This can be argued but this is how I motivated it when I was asked about it during the dask workshop

that are homogeneous in size with many of the clusters dynamically scaled. The peculiarities of this
setup have in some cases triggered edge cases and uncovered bugs, which lead us to submit a number
of upstream pull requests.

### Cluster separation

> We are running separate Dask clusters for separate customers in order to isolate the customers from each other.

Reviewer 2: What is a separate customer? ;-)

```suggestion
To provide isolatation between customer environments we run separate Dask clusters per customer.
```

For the same reason, we also run separate clusters for production, staging, and development environments. 

> But it does not stop there. Our compute pipeline involves different steps, for instance, downloading the input data

Reviewer 2: ```suggestion
But it does not stop there. Our compute pipeline involves different steps, for instance downloading the input data
```

unless there is some comma rule that I don't know about ;-)

Reviewer 4: This section describes what I referred to as environment isolation.
I would motivate it with having different products which build onto each others and are running sequentially. Since every product / application operates mostly autonomously, we separated the environments off each others. We might also drop a note about having different runtime environnments (python only, C++, lately even java, etc.)

Reviewer 2: But that's the **rest of this paragraph**, that you're talking about.
This sentence/part really focuses on what the pipeline is comprised of and therefore is not really Cluster separation?!

So IMHO, this sections does not convey a clear message and as a result there is a disconnect between heading and content.

Author: I wanted to motivate why we are using different clusters for different steps of the pipeline; I think that's what's special about us. I'll make this more clear.

For performance reasons, we install the Python packages holding the code needed
 for the computations on each worker. Since we do not want to synchronize the Python packages required for each pipeline
 step (e.g., allow different Pandas versions for calculating order proposals and predictions in order not to introduce
 a coupling of the software release cycles), we have to run a separate Dask cluster for each of pipeline steps. 
This means we operate more than
ten Dask clusters per customer and environment, with most of the time, only one of the clusters being active
and computing something. While this leads to overhead in terms of administration and hardware resouces,
it also gives us a lot of flexibility. For instance, we can update the software on the cluster of one part of the compute pipeline
while another part of the pipeline is computing something on a different cluster.

### Some numbers

The number and size of the workers varies from cluster to cluster depending on the degree of parallelism
of the computation being performed, its resource requirements, and the available timeframe for the computation.

> At the time of writing this, we are running more than 500 distinct clusters.

Reviewer 2: `At the time of writing` is a fixed term. You should omit the "this"

```suggestion
At the time of writing, we are running more than 500 distinct clusters.
```

> They have between one and 225 workers, with worker size varying between 1GiB and 64GiB of memory.

Reviewer 2: You should IMHO preface, that/why we only care about memory and not about storage/cores?!

> Our total memory use goes up to as much as 15TiB.

Reviewer 2: Total of what? One cluster or all clusters?

> Our total memory use goes up to as much as 15TiB.

Reviewer 4: how many CPUs? 

Can we visualise this somehow (if there is no sane visualization, that's also fine, of course)?

## Cluster scaling and resilience

To improve manageability, resilience, and resource utilization, we run the Dask clusters on top
of [Apache Mesos](http://mesos.apache.org/)/[Aurora](http://aurora.apache.org/) 
and [Kubernetes](https://kubernetes.io/). This means every worker as well as the scheduler and client

> each run in an insolated container. Communication happens via reverse proxies to make the communication

Reviewer 4: ```suggestion
each run in an isolated container. Communication happens via reverse proxies to make the communication
```

unless you're making a thermodynamical point I didn't get ;)

to the load on the cluster. This is possible since Mesos/Kubernetes . This
makes it really easy to add or remove worker instances from an existing Dask cluster.

To determine the optimum number of worker instances to run, we added the ``desired_workers`` metric to Distributed.
The metric exposes
the degree of parallelism that a computation has and thus allows us to infer how much workers a cluster should
ideally have. Based on this metric, as well as on the overall resources available and on fairness criteria
(remember, we run a lot of Dask/Distributed clusters), we add or remove workers to our clusters.
To resolve the problem of balancing the conflicting requirements for different resouces like RAM or CPUs
 using [Dominant Resource Fairness](https://cs.stanford.edu/~matei/papers/2011/nsdi_drf.pdf).

## Dask issues

The particular way we use Dask, especially running it in containers connected by reverse proxies and the fact
that we dynamically add/remove workers from a cluster quite frequently for autoscaling has lead us to hit some
edge cases and instabilities and given us the chance to contribute some fixes and improvements to Dask.

> For instance, we were able to 

Reviewer 4: By now, the list of fixes we contributed is quite long. Shall we put in more of these?

Author: I'm not sure if we should really put a long list in here, it could look a bit like showing off. I'd just list the top ones, not sure if I got the right ones, though.

[improve stability after connection failures](https://github.com/dask/distributed/pull/3246) or when 
[workers are removed from the cluster](https://github.com/dask/distributed/pull/3366). 

If you are interested in our contributions to Dask and our commitment to the dask community, please

> also check out our blog post [Karlsruhe to D.C. ― a Dask story](2020-03-24-dask-developer-workshop.markdown).

Reviewer: ```suggestion
also check out our blog post [Karlsruhe to D.C. ― a Dask story](2020-05-28-dask-developer-workshop.markdown).
```
We just changed the date before publishing. You may want to rebase and check using the jekyll commands
